protocols/relay/src/v2/client.rs

```diff
@@ -146,11 +152,40 @@ impl Client {
 }
 
 impl NetworkBehaviour for Client {
-    type ConnectionHandler = handler::Prototype;
+    type ConnectionHandler = Either<Handler, dummy::ConnectionHandler>;
     type OutEvent = Event;
 
-    fn new_handler(&mut self) -> Self::ConnectionHandler {
-        handler::Prototype::new(self.local_peer_id, None)
+    fn new_handler(
+        &mut self,
+        peer: &PeerId,
+        connected_point: &ConnectedPoint,
+    ) -> Result<Self::ConnectionHandler, Box<dyn std::error::Error + Send + 'static>> {
+        if connected_point.is_relayed() {
+            if let Some(event) = self.initial_events.remove(peer) {
+                log::debug!(
+                    "Established relayed instead of direct connection to {:?}, \
+                     dropping initial in event {:?}.",
+                    peer,
+                    event
+                );
+            }
+
+            // Deny all substreams on relayed connection.
+            Ok(Either::Right(dummy::ConnectionHandler))
+        } else {
+            let mut handler = Handler::new(
+                self.local_peer_id,
+                *peer,
+                connected_point.get_remote_address().clone(),
+            );
+
+            if let Some(event) = self.initial_events.remove(peer) {
+                #[allow(deprecated)]
+                handler.inject_event(event)
+            }
```

Comment:
Injecting events through two ways, i.e. (1) at creation and (2) regularly during the lifetime of the handler via `Swarm`, seems inconsistent to me. That said, I can not think of a better way. I suggest keeping as is.

Reply:
We could hold the event back here and instead listen for the `ConnectionEstablished` event and then emit an event later?